add analyzer for virtualservice overlapping matches. Fixed issue#24707

zhlsunshine opened the pull request:
This pr fix the overlapping matches for virtualservice, please refer [issue#24707](https://github.com/istio/istio/issues/24707) for more details



[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

zhlsunshine commented:
Here is the test result:

create the wrong virtualservice(yaml):
```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: reviews-route
spec:
  hosts:
  - reviews
  http:
  - name: "reviews-v2-routes"
    match:
    - uri:
        prefix: "/wpcatalog"
    - uri:
        prefix: "/consumercatalog"
    rewrite:
      uri: "/newcatalog"
    route:
    - destination:
        host: reviews
        subset: v2
  - name: "reviews-v1-route"
    match:
    - uri:
        prefix: "/wpcatalogs"
    route:
    - destination:
        host: reviews
        subset: v1
```

```
$istioctl analyze -A
Error [IST0101] (VirtualService reviews-route.default) Referenced host+subset in destinationrule not found: "reviews+v1"
Error [IST0101] (VirtualService reviews-route.default) Referenced host+subset in destinationrule not found: "reviews+v2"
Info [IST0102] (Namespace calico-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace calico-system istio-injection=enabled' to enable it, or 'kubectl label namespace calico-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace istio-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace istio-system istio-injection=enabled' to enable it, or 'kubectl label namespace istio-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace tigera-operator) The namespace is not enabled for Istio injection. Run 'kubectl label namespace tigera-operator istio-injection=enabled' to enable it, or 'kubectl label namespace tigera-operator istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0118] (Service calico-typha.calico-system) Port name calico-typha (port: 5473, targetPort: calico-typha) doesn't follow the naming convention of Istio port.
Info [IST0131] (VirtualService reviews-route.default) VirtualService rule "reviews-v1-route" match #0 of prefix /wpcatalogs is not used (duplicates/overlapping a match in rule #0 of prefix /wpcatalog on "reviews-v2-routes").
Error: Analyzers found issues when analyzing all namespaces.
```

istio-testing commented:
@zhlsunshine: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | c4b8242fe850cd53d435074a138ef5804b5b5a89 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/release-notes_istio/1390538189927616512) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@zhlsunshine: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | c4b8242fe850cd53d435074a138ef5804b5b5a89 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/release-notes_istio/1390538189927616512) | `/test release-notes_istio`
unit-tests_istio | c4b8242fe850cd53d435074a138ef5804b5b5a89 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/unit-tests_istio/1390538183237701632) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

zhlsunshine commented:
/test release-notes_istio

istio-testing commented:
@zhlsunshine: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 96cd7e10dd1f87ecbddae8217aa169cc2498c1af | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/unit-tests_istio/1390555076002582528) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

zhlsunshine commented:
/test unit-tests_istio

istio-testing commented:
@zhlsunshine: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 96cd7e10dd1f87ecbddae8217aa169cc2498c1af | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/unit-tests_istio/1390574235931381760) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@zhlsunshine: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | f5dac2688142d41a3472931b7d0558066566944a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/unit-tests_istio/1390586317166874624) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@zhlsunshine: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | f5dac2688142d41a3472931b7d0558066566944a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/unit-tests_istio/1390586317166874624) | `/test unit-tests_istio`
integ-pilot-multicluster-tests_istio | f5dac2688142d41a3472931b7d0558066566944a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/integ-pilot-multicluster-tests_istio/1390586318844596224) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@zhlsunshine: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-multicluster-tests_istio | f5dac2688142d41a3472931b7d0558066566944a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/integ-pilot-multicluster-tests_istio/1390586318844596224) | `/test integ-pilot-multicluster-tests_istio`
unit-tests_istio | 8c5cb2f80c6259f0065477c5bcd2f1d4600d56bb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/unit-tests_istio/1390597776236089344) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@zhlsunshine: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-multicluster-tests_istio | 959ce3bfdc88dc103ab678d26d8962603015b0fb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/integ-pilot-multicluster-tests_istio/1390601579907780608) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

shamsher31 commented on galley/pkg/config/analysis/msg/messages.yaml:
```diff
@@ -317,7 +317,7 @@ messages:
     level: Info
     description: "A VirtualService rule match duplicates a match in a previous rule."
-    template: "VirtualService rule %v match %v is not used (duplicates a match in rule %v)."
+    template: "VirtualService rule %v match %v is not used (duplicates/overlapping a match in rule %v)."
```
```suggestion
    template: "VirtualService rule %v match %v is not used (duplicate/overlapping match in rule %v)."
```

zhlsunshine replied:
got it!

zhlsunshine commented:
Hi @shamsher31, here is the new output below:

```
$istioctl analyze -A
......
......
Info [IST0131] (VirtualService reviews-route.default) VirtualService rule "reviews-v1-route" match #0 of prefix /wpcatalogs is not used (duplicate/overlapping match in rule #0 of prefix /wpcatalog on "reviews-v2-routes").
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.10/docs/reference/config/analysis for more information about causes and resolutions.
```

shamsher31 reviewed (commented):
Overall LGTM. I will let @esnible take a another look as he wrote the previous analyzer.

esnible commented:
Please write test cases and put them in _galley/pkg/config/analysis/analyzers/analyzers_test.go_ under the "dupmatches" test.

Create a file similar to _virtualservice_dupmatches.yaml_ which includes this new kind of overlapping match.  Write test cases.  Difficult and evil test cases are needed.

This VirtualService has no overlapping matches even though it has identical URI prefixes.  I did not try your code, but I suspect it will flag incorrectly this as having a problem.

```
kubectl apply -f - <<EOF
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: duplicate-match
spec:
  hosts:
  - sample.bar.svc.cluster.local
  http:
  - name: send GET /api/v1/products to sample.foo
    match:
    - uri:
        prefix: /api/v1/products
      method:
        exact: GET
    route:
    - destination:
        host: sample.foo.svc.cluster.local
  - name: send POST /api/v1/products to sample.bar
    match:
    - name: prefix /api/v1/products
      method:
        exact: POST
      uri:
        prefix: /api/v1/products
    route:
    - destination:
        host: sample.bar.svc.cluster.local
EOF
```

esnible reviewed (changes requested):
Please write test cases and put them in _galley/pkg/config/analysis/analyzers/analyzers_test.go_ under the "dupmatches" test.

Create a file similar to _virtualservice_dupmatches.yaml_ which includes this new kind of overlapping match.  Write test cases.  Difficult and evil test cases are needed.

This VirtualService has no overlapping matches even though it has identical URI prefixes.  I did not try your code, but I suspect it will flag incorrectly this as having a problem.

```
kubectl apply -f - <<EOF
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: duplicate-match
spec:
  hosts:
  - sample.bar.svc.cluster.local
  http:
  - name: send GET /api/v1/products to sample.foo
    match:
    - uri:
        prefix: /api/v1/products
      method:
        exact: GET
    route:
    - destination:
        host: sample.foo.svc.cluster.local
  - name: send POST /api/v1/products to sample.bar
    match:
    - name: prefix /api/v1/products
      method:
        exact: POST
      uri:
        prefix: /api/v1/products
    route:
    - destination:
        host: sample.bar.svc.cluster.local
EOF
```

zhlsunshine commented:
> Please write test cases and put them in _galley/pkg/config/analysis/analyzers/analyzers_test.go_ under the "dupmatches" test.
> 
> Create a file similar to _virtualservice_dupmatches.yaml_ which includes this new kind of overlapping match. Write test cases. Difficult and evil test cases are needed.
> 
> This VirtualService has no overlapping matches even though it has identical URI prefixes. I did not try your code, but I suspect it will flag incorrectly this as having a problem.
> 
> ```
> kubectl apply -f - <<EOF
> apiVersion: networking.istio.io/v1alpha3
> kind: VirtualService
> metadata:
>   name: duplicate-match
> spec:
>   hosts:
>   - sample.bar.svc.cluster.local
>   http:
>   - name: send GET /api/v1/products to sample.foo
>     match:
>     - uri:
>         prefix: /api/v1/products
>       method:
>         exact: GET
>     route:
>     - destination:
>         host: sample.foo.svc.cluster.local
>   - name: send POST /api/v1/products to sample.bar
>     match:
>     - name: prefix /api/v1/products
>       method:
>         exact: POST
>       uri:
>         prefix: /api/v1/products
>     route:
>     - destination:
>         host: sample.bar.svc.cluster.local
> EOF
> ```

------

Hi @esnible, I understand that you have mentioned some in issue# , u r correct, there would be a misanalysis for my previous submit . So I updated it and it's correct as below. And the test case would be included in the newest submit.

```
$ kubectl apply -f - <<EOF
> apiVersion: networking.istio.io/v1alpha3
> kind: VirtualService
> metadata:
>   name: duplicate-match
> spec:
>   hosts:
>   - sample.bar.svc.cluster.local
>   http:
>   - name: send GET /api/v1/products to sample.foo
>     match:
>     - uri:
>         prefix: /api/v1/products
>       method:
>         exact: GET
>     route:
>     - destination:
>         host: sample.foo.svc.cluster.local
>   - name: send POST /api/v1/products to sample.bar
>     match:
>     - name: prefix /api/v1/products
>       method:
>         exact: POST
>       uri:
>         prefix: /api/v1/products
>     route:
>     - destination:
>         host: sample.bar.svc.cluster.local
> EOF
virtualservice.networking.istio.io/duplicate-match created
```
And 
```
$istioctl analyze -A
......
......
Info [IST0131] (VirtualService reviews-route.default) VirtualService rule "reviews-v1-route" match #0 of prefix /wpcatalogs is not used (duplicate/overlapping match in rule #0 of prefix /wpcatalog on "reviews-v2-routes").
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.10/docs/reference/config/analysis for more information about causes and resolutions.
```

istio-testing commented:
@zhlsunshine: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3f1f9cdca47fc2c5b1b5225a3e1a6d11db68efb5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/integ-multicluster-k8s-tests_istio/1396737980781039617) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@zhlsunshine: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3f1f9cdca47fc2c5b1b5225a3e1a6d11db68efb5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/integ-multicluster-k8s-tests_istio/1396737980781039617) | `/test integ-multicluster-k8s-tests_istio`
integ-pilot-k8s-tests_istio | 3f1f9cdca47fc2c5b1b5225a3e1a6d11db68efb5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/integ-pilot-k8s-tests_istio/1396737980617461760) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

zhlsunshine commented:
/test integ-multicluster-k8s-tests_istio

zhlsunshine commented:
/test integ-pilot-k8s-tests_istio

zhlsunshine commented:
Hi @shamsher31 & @esnible,  Please help to review my new submit again when u r available, I have fixed bug above mentioned by @esnible and added necessary test case under the "dupmatches" test. And below is my test result. Thank you very much for your time! ^_^.  
```
$istioctl analyze -A
......
......
Info [IST0131] (VirtualService overlapping-in-single-match.istio-system) VirtualService rule "send product to sample.foo" match #1 of prefix /api/v1/products is not used (duplicate/overlapping match in rule #0 of prefix /api/v1/product on "send product to sample.foo").
Info [IST0131] (VirtualService overlapping-in-two-matches.istio-system) VirtualService rule "send products to sample.bar" match #0 of prefix /api/v1/products is not used (duplicate/overlapping match in rule #0 of prefix /api/v1/product on "send product to sample.foo").
Info [IST0131] (VirtualService overlapping-mathes-with-different-methods.default) VirtualService rule "send products to sample.bar" match #0 of prefix /api/v1/product is not used (duplicate/overlapping match in rule #0 of prefix /api/v1/prod on "send product to sample.foo").
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.11/docs/reference/config/analysis for more information about causes and resolutions.
```

esnible reviewed (commented):
This is much better!  Please keep working on it.

I would like more tests.  Merely testing Method is not enough because you have added logic for other fields.  Include cases where a field is unset.

I want to see
- a case where rule 1 is `GET /` and rule 2 is `<unset-method> /`.
- a case where rule 1 is prefix=`/` and rule 2 is exact=`/`.

The function analyzeUnreachableHTTPRules() is getting large.  Perhaps refactor out the comparison of matches.

Although this PR catches more overlap, it doesn't catch everything.  For example, if the first match is `GET / ; headers=Content-Type: text/html` and the second is `GET / ; headers=Content-Type: text/html,Content-Encoding: gzip`.

Converting the other match params into JSON and concatenating them is strange.  It may work, but perhaps there is a better way?  Perhaps break the problem down formally.  Perhaps

- A rule X is completely _covered_ by rule Y if
   - X.Method completely covers Y.Method &&
   - X.Authority completely covers Y.Authority &&
   - X.Headers completely covers Y.Headers &&
   - X.Port == Y.Port &&
   - X.QueryParams completely covers Y.QueryParams &&
   - X.WithoutHeaders completely covers Y.WithoutHeaders
- A rule N is unreachable if any rule [0..N-1] _covers_ it.

We need logic to check 'cover' for two Map types, and for two StringMatch.  The StringMatch test can ignore regexps.

zhlsunshine commented:
> This is much better! Please keep working on it.
> 
> I would like more tests. Merely testing Method is not enough because you have added logic for other fields. Include cases where a field is unset.
> 
> I want to see
> 
>     * a case where rule 1 is `GET /` and rule 2 is `<unset-method> /`.
> 
>     * a case where rule 1 is prefix=`/` and rule 2 is exact=`/`.
> 
> 
> The function analyzeUnreachableHTTPRules() is getting large. Perhaps refactor out the comparison of matches.
> 
> Although this PR catches more overlap, it doesn't catch everything. For example, if the first match is `GET / ; headers=Content-Type: text/html` and the second is `GET / ; headers=Content-Type: text/html,Content-Encoding: gzip`.
> 
> Converting the other match params into JSON and concatenating them is strange. It may work, but perhaps there is a better way? Perhaps break the problem down formally. Perhaps
> 
>     * A rule X is completely _covered_ by rule Y if
>       
>       * X.Method completely covers Y.Method &&
>       * X.Authority completely covers Y.Authority &&
>       * X.Headers completely covers Y.Headers &&
>       * X.Port == Y.Port &&
>       * X.QueryParams completely covers Y.QueryParams &&
>       * X.WithoutHeaders completely covers Y.WithoutHeaders
> 
>     * A rule N is unreachable if any rule [0..N-1] _covers_ it.
> 
> 
> We need logic to check 'cover' for two Map types, and for two StringMatch. The StringMatch test can ignore regexps.

@esnible okay, I will add these 2 test cases.  And I understand your worries on growing function analyzeUnreachableHTTPRules(), and it's for re-using `route.Match` loop to retrieve information. And thank you for your advices on covered rules above, however there may be more to discuss on such rules. It's complicated as you mentioned earlier,  and I think it's reasonable to import some rule assertions to implement it, for instance such [hostname match](https://github.com/istio/istio/blob/ef9bdaa405fb346acd6132e78052a12744b0a7fe/pkg/config/host/name.go#L36), then I think it would be a great time to refactor the funciton analyzeUnreachableHTTPRules, is that okay?

zhlsunshine commented:
Hi @esnible , I think we may need to discuss what exactly `completely covered` means, I understand that it should be similar with [SubsetOf](https://github.com/istio/istio/blob/ef9bdaa405fb346acd6132e78052a12744b0a7fe/pkg/config/host/name.go#L63). If yes, I think there should be no any `asJSON` used in PR.

zhlsunshine commented:
Hi @esnible & @shamsher31 , according to above comments, I changed code to implement the covered rules mentioned by @esnible, and I also added 2 cases in _galley/pkg/config/analysis/analyzers/testdata/virtualservice_overlappingmatches.yaml_:

- a case where rule 1 is GET / and rule 2 is <unset-method> /.
- a case where rule 1 is prefix=`/` and rule 2 is exact=`/`.

I have handled the first one, and the rule with method GET and another rule with unset-method is the same thanks to the unset-method should have the default behavior as GET. And below is the test result:

```
$istioctl analyze -A
......
......
Info [IST0131] (VirtualService non-method-get.default) VirtualService rule "send product to sample.foo" match #1 of prefix /api/v1/products is not used (duplicate/overlapping match in rule #0 of prefix /api/v1/product on "send product to sample.foo").
Info [IST0131] (VirtualService overlapping-in-single-match.default) VirtualService rule "send product to sample.foo" match #1 of prefix /api/v1/products is not used (duplicate/overlapping match in rule #0 of prefix /api/v1/product on "send product to sample.foo").
Info [IST0131] (VirtualService overlapping-in-two-matches.default) VirtualService rule "send products to sample.bar" match #0 of prefix /api/v1/products is not used (duplicate/overlapping match in rule #0 of prefix /api/v1/product on "send product to sample.foo").
Info [IST0131] (VirtualService overlapping-mathes-with-different-methods.default) VirtualService rule "send products to sample.bar" match #0 of prefix /api/v1/product is not used (duplicate/overlapping match in rule #0 of prefix /api/v1/prod on "send product to sample.foo").
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.11/docs/reference/config/analysis for more information about causes and resolutions.
```

zhlsunshine committed:
according to the comments in this PR, implemented the cover rules for overlapping match and added 2 test cases

zhlsunshine committed:
according to the comments in this PR, implemented the cover rules for overlapping match and added 2 test cases

shamsher31 commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,171 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
+func coveredValidation(vA, vB *OverlappingForHTTPRoute) bool {
+	// check the URI overlapping match, such as vB.Prefix is '/debugs' and vA.Prefix is '/debug'
+	if strings.HasPrefix(vB.Prefix, vA.Prefix) {
+		// checke the port field
```
```suggestion
		// check the port field
```

zhlsunshine replied:
done

shamsher31 commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,171 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
+			return false
+		}
+
+		// checke the match method
```
```suggestion
		// check the match method
```

zhlsunshine replied:
done

shamsher31 commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,171 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
+			}
+		}
+
+		// checke the match authority
```
```suggestion
		// check the match authority
```

zhlsunshine replied:
done

shamsher31 commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,171 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
+			}
+		}
+
+		// checke the match Headers
```
```suggestion
		// check the match Headers
```

zhlsunshine replied:
done

shamsher31 commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,171 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
+			}
+		}
+
+		// checke the match QueryParams
```
```suggestion
		// check the match QueryParams
```

zhlsunshine replied:
done

shamsher31 commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,171 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
+			}
+		}
+
+		// checke the match NonHeaders
```
```suggestion
		// check the match NonHeaders
```

zhlsunshine replied:
Okay, I will update it

zhlsunshine replied:
done!

shamsher31 reviewed (approved):
Minor nit, otherwise LGTM.

hzxuzhonghu reviewed (commented):
I would expect the ut for validation. Actually with validating webhook, invalid vs will be rejected ealier

istio-testing commented:
@zhlsunshine: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-telemetry-k8s-tests_istio | 7aef547d37ec1dba90040c082b0d3309d442c6bf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/32701/integ-telemetry-k8s-tests_istio/1403266178054361088) | `/test integ-telemetry-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

zhlsunshine commented:
/test integ-telemetry-k8s-tests_istio

zhlsunshine commented:
> I would expect the ut for validation. Actually with validating webhook, invalid vs will be rejected ealier

okay, I will add unit test for this new analysis

zhlsunshine committed:
change some grammatical errors

zhlsunshine committed:
add more commentaries for this PR change according to the reviewer

hzxuzhonghu commented on pkg/config/validation/validation.go:
```diff
@@ -130,6 +133,18 @@ type AnalysisAwareError struct {
 	Parameters []interface{}
 }
 
+type OverlappingForHTTPRoute struct {
```
please add some comments on this structure

zhlsunshine replied:
it's done

hzxuzhonghu commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,175 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
 		return errs.Unwrap()
 	})
 
+func assignExactOrPrefix(exact, prefix string) string {
```
it  seems exact and prefix can not be set at the same time, so the if/ else clause can be simplified

zhlsunshine replied:
it's done

hzxuzhonghu commented on pkg/config/validation/validation.go:
```diff
@@ -1960,10 +1975,175 @@ var ValidateVirtualService = registerValidateFunc("ValidateVirtualService",
+			vBhdValue, ok := vB.MatchHeaders[hdKey]
+			if !ok {
+				return false
+			} else if strings.Compare(hdValue, vBhdValue) != 0 {
```
nit: hdValue!=hdValue

zhlsunshine replied:
it's done

hzxuzhonghu reviewed (approved):
Left some nits, can be addressed as a followup

zhlsunshine commented:
> Left some nits, can be addressed as a followup

@hzxuzhonghu I have completed the comments, but  still focusing on development of unit tests , I will go on focusing on it, thanks!
